src/pages/index.js

```diff
@@ -30,72 +32,219 @@ const previewText = previewImageModal.querySelector(".modal__preview-text");
 const addFormElement = document.querySelector("#add-form");
+const avatarFormelement = document.querySelector("#avatar-form");
 
 const editFormValidator = new FormValidator(config, editFormElement);
 const addFormValidator = new FormValidator(config, addFormElement);
+const avatarFormValidator = new FormValidator(config, avatarFormelement);
 
 editFormValidator.enableValidation();
 addFormValidator.enableValidation();
+avatarFormValidator.enableValidation();
 
-// form popup
-const newFormPopup = new PopupWithForm("#card-modal", handleFormSubmit);
-newFormPopup.setEventListeners();
+// Card Form Popup //
 
-//  profile edit popup
-const userInfo = new UserInfo(".profile__title", ".profile__text");
-const profileEditPopup = new PopupWithForm("#edit-modal", (data) => {
-  userInfo.setUserInfo(data);
-  profileEditPopup.close();
-});
+const newCardPopup = new PopupWithForm(
+  "#card-modal",
+  handleCardFormSubmit,
+  "Create"
+);
+newCardPopup.setEventListeners();
+
+//  Profile Edit Popup //
+
+const userInfo = new UserInfo(
+  ".profile__title",
+  ".profile__text",
+  ".profile__image"
+);
+
+const profileEditPopup = new PopupWithForm(
+  "#edit-modal",
+  handleProfileEdit,
+  "Save"
+);
 profileEditPopup.setEventListeners();
 
-// image popup
+// Image Popup //
+
 const imagePopup = new PopupWithimage("#image-modal");
 imagePopup.setEventListeners();
 
-// section class
-const section = new Section(
-  {
-    items: initialCards,
-    renderer: (item) => {
-      const cardEl = renderCard(item);
-      section.addItem(cardEl);
-    },
-  },
-  ".cards__list"
+// Avatar Popup //
+
+const avatarPopup = new PopupWithForm(
+  "#avatar-modal",
+  handleAvatarFormSubmit,
+  "Save"
+);
+avatarPopup.setEventListeners();
+
+// Delete Card Popup //
+
+const deleteCardPopup = new PopupWithConfirmation(
+  "#confirm-modal",
+  "Deleting..."
 );
-section.renderItems();
+deleteCardPopup.setEventListeners();
+
+// functions //
 
-// functions
 function renderCard(data) {
-  const card = new Card(data, "#card-template", handleImageClick);
-  return card.getView();
+  const cardEl = new Card(
+    data,
+    "#card-template",
+    handleImageClick,
+    handleDeleteclick,
+    handleLikeClick
+  );
+  section.addItem(cardEl.getView());
+}
+
+function handleCardFormSubmit(data) {
+  newCardPopup.setLoading(true);
+  api
+    .addCard(data)
+    .then((data) => {
+      renderCard(data);
+    })
+    .then(() => {
+      newCardPopup.close();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      newCardPopup.setLoading(false, "Create");
+    });
+}
+
+function handleProfileEdit(data) {
+  profileEditPopup.setLoading(true);
+  api
+    .editUserInfo(data)
+    .then((data) => {
+      userInfo.setUserInfo(data);
+      profileEditPopup.close();
+    })
+    .catch((err) => {
+      console.error(err);
+    })
+    .finally(() => {
+      profileEditPopup.setLoading(false);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```